### PR TITLE
Sort, constant, mod process

### PR DESCRIPTION
**Pull request by zansinergise**
https://git.sinergise.com/team-6/openeo-platform/-/issues/82

### CODE REVIEW

File: src/pg_to_evalscript/javascript_processes/mod.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  if (typeof x !== "number") {
+    throw new Error("Argument `x` is not a number.");
+  }
+
+  if (typeof y !== "number") {
+    throw new Error("Argument `y` is not a number.");
+  }
+
+  return x % y;
```
Comment (zcernigoj) [inline src/pg_to_evalscript/javascript_processes/mod.js:28]: From the [docs](https://docs.openeo.cloud/processes/#mod)
> The result of a modulo operation has the sign of the divisor.

Be sure to add the sign of the divisor (`y`) to the result. 

For example, `3 % -2` returns `1` in javascript.

File: src/pg_to_evalscript/javascript_processes/sort.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+function sort(arguments) {
+  const { data, asc = true, nodata = null } = arguments;
+  let newData = [...data];
+  let arr_of_nulls = newData.filter((x) => x === null);
+
+  newData = newData.filter((x) => x !== null);
```
Comment (zcernigoj) [inline src/pg_to_evalscript/javascript_processes/sort.js:6]: This can be refactored to

```javascript
let newData = [...data].filter((x) => x !== null);
let arr_of_nulls = [...data].filter((x) => x === null);
```

I also think a check if `data` is array and if `asc` and `nodata` are either boolean or `null` should be added at the beginning.

File: src/pg_to_evalscript/javascript_processes/sort.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  if (nodata === false) {
+    newData = [...arr_of_nulls, ...newData];
+  }
+
+  return newData;
+}
+
+function compareFn(x, y) {
+  const isValidDate = function (date) {
+    return new Date(date) !== "Invalid Date" && !isNaN(new Date(date));
```
Comment (zcernigoj) [inline src/pg_to_evalscript/javascript_processes/sort.js:27]: I experimented with `new Date(date) !== "Invalid Date"` and it always returns `true`.

To make this part useful, the `dateObject.toString()` should be used - this returns `Invalid date` for the invalid dates.

File: src/pg_to_evalscript/javascript_processes/sort.js (hunk 2 above)
Comment (zcernigoj) [reply src/pg_to_evalscript/javascript_processes/sort.js:6]: Maybe there should also be a check for every element of the array (can be number, string or `null`, but not boolean, array, etc.)

File: tests/unit_tests/test_mod.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+@pytest.fixture
+def mod_process_code():
+    return load_process_code("mod")
+
+
+@pytest.mark.parametrize(
+    "example_input,expected_output",
+    [
+        ({"x": 27, "y": 5}, 2),
+        ({"x": -27, "y": 5}, -2),
```
Comment (zcernigoj) [inline tests/unit_tests/test_mod.py:17]: Result should be positive

File: tests/unit_tests/test_mod.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        ({"x": 3.14, "y": -2}, 1.14),
```
Comment (zcernigoj) [inline tests/unit_tests/test_mod.py:18]: Result should be negative

File: tests/unit_tests/test_mod.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        ({"x": -27, "y": -5}, -2),
+        ({"x": -3, "y": -1}, 0),
+        ({"x": 3, "y": -2}, 1),
```
Comment (zcernigoj) [inline tests/unit_tests/test_mod.py:21]: Result should be negative

File: tests/unit_tests/test_mod.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        ({"x": 3, "y": 4}, 3),
+        ({"x": None, "y": 3}, None),
+        ({"x": 1, "y": None}, None),
```
Comment (zcernigoj) [inline tests/unit_tests/test_mod.py:24]: please add test where both are `None`.

File: tests/unit_tests/test_sort.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+@pytest.fixture
+def sort_process_code():
+    return load_process_code("sort")
+
+
+@pytest.mark.parametrize(
+    "example_input,expected_output",
+    [
+        ({"data": [6, -1, 2, None, 7, 4, None, 8, 3, 9, 9]}, [-1, 2, 3, 4, 6, 7, 8, 9, 9]),
```
Comment (zcernigoj) [inline tests/unit_tests/test_sort.py:16]: Please add test where all elements are `None`.

**Review by zcernigoj — CHANGES_REQUESTED** [review]
`mod` and `sort` need improvements (comments above)

File: src/pg_to_evalscript/javascript_processes/mod.js (hunk 1 above)
Comment (zcernigoj) [reply src/pg_to_evalscript/javascript_processes/mod.js:28]: Added comments in the `test_mod.py` for convenience

File: src/pg_to_evalscript/javascript_processes/sort.js (hunk 2 above)
Comment (zcernigoj) [reply src/pg_to_evalscript/javascript_processes/sort.js:6]: Optimization to go through the array less times (once instead of 2 or 3 times)

- go through the array with the `for` loop
- check if the element is of appropriate type
- check if it's `null`
  - if it is, add it to `arr_of_nulls`
  - else, add it to the `newData`

File: tests/unit_tests/test_mod.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+
+@pytest.mark.parametrize(
+    "example_input,expected_output",
+    [
+        ({"x": 27, "y": 5}, 2),
+        ({"x": -27, "y": 5}, -2),
+        ({"x": 3.14, "y": -2}, 1.14),
+        ({"x": -27, "y": -5}, -2),
+        ({"x": -3, "y": -1}, 0),
```
[the remark this answers is not in the document]
Comment (zansinergise) [reply tests/unit_tests/test_mod.py:20]: I guess JS/python does not differ between `+0` and `-0` since in both cases the test case passes

File: tests/unit_tests/test_mod.py (hunk 7 above)
Comment (zansinergise) [reply tests/unit_tests/test_mod.py:24]: Added

File: tests/unit_tests/test_mod.py (hunk 6 above)
Comment (zansinergise) [reply tests/unit_tests/test_mod.py:21]: Fixed

File: tests/unit_tests/test_mod.py (hunk 5 above)
Comment (zansinergise) [reply tests/unit_tests/test_mod.py:18]: Fixed

File: tests/unit_tests/test_mod.py (hunk 4 above)
Comment (zansinergise) [reply tests/unit_tests/test_mod.py:17]: Fixed

File: src/pg_to_evalscript/javascript_processes/mod.js (hunk 1 above)
Comment (zansinergise) [reply src/pg_to_evalscript/javascript_processes/mod.js:28]: Should be fixed

File: src/pg_to_evalscript/javascript_processes/sort.js (hunk 2 above)
Comment (zansinergise) [reply src/pg_to_evalscript/javascript_processes/sort.js:6]: Should be fixed.

File: src/pg_to_evalscript/javascript_processes/sort.js (hunk 3 above)
Comment (zansinergise) [reply src/pg_to_evalscript/javascript_processes/sort.js:27]: Should be fixed now

File: tests/unit_tests/test_sort.py (hunk 8 above)
Comment (zansinergise) [reply tests/unit_tests/test_sort.py:16]: Added

File: tests/unit_tests/test_mod.py (hunk 7 above)
Comment (zcernigoj) [reply tests/unit_tests/test_mod.py:24]: Thanks!

File: src/pg_to_evalscript/javascript_processes/sort.js (hunk 2 above)
Comment (zcernigoj) [reply src/pg_to_evalscript/javascript_processes/sort.js]: Thanks, looks good.

File: tests/unit_tests/test_sort.py (hunk 8 above)
Comment (zcernigoj) [reply tests/unit_tests/test_sort.py:16]: Thanks!

File: tests/unit_tests/test_mod.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zcernigoj) [reply tests/unit_tests/test_mod.py]: Thanks for checking.

**Review by zcernigoj — APPROVED** [review]
